src/genetic_algorithm/algorithms/species_based_algorithm.rs

```diff
@@ -0,0 +1,176 @@
+                }
+            }
+
+            if species.len() % 2 != 0 {
+                let first_parent = &species.last().unwrap();
+                let second_parent = population[species.len() - 1];
+                let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                first_child = first_child.mutate(1.0);
+                second_child = second_child.mutate(1.0);
+                first_child.correct();
+                second_child.correct();
+                offspring_population.push(first_child);
+                offspring_population.push(second_child);
+            }
+        }
+
+        offspring_population.truncate(self.population.values().flatten().count());
+        self.offspring_population = offspring_population;
+
+        dbg!(&self.offspring_population.len());
+        dbg!(&self.population.values().flatten().count());
+    }
+
+    fn replacement(&mut self) {
+        self.population.clear();
+        for chromosome in self.new_population.iter() {
+            self.population
+                .entry(chromosome.characteristic())
+                .or_insert(Vec::new())
+                .push(chromosome.clone());
+        }
+        self.new_population.clear();
+    }
+
+    fn finished_evaluating(&mut self, chromosome: T) {
+        self.new_population.push(chromosome);
+    }
+
+    fn all_have_finished_evaluating(&self, population_size: usize) -> bool {
+        self.new_population.len() == population_size * 2
+    }
+
+    fn save_results(&self, generation_count: usize) {
+        let mut population: Vec<&T> = self.population.values().flatten().clone().collect();
+        population.sort_by(|a, b| b.get_fitness().partial_cmp(&a.get_fitness()).unwrap());
+
+        let pretty_config = PrettyConfig::default();
+        let best = to_string_pretty(
+            population
+                .iter()
+                .max_by(|x, y| x.get_fitness().partial_cmp(&y.get_fitness()).unwrap())
+                .unwrap(),
+            pretty_config.clone(),
+        )
+        .unwrap();
+        let worst = to_string_pretty(
+            population
+                .iter()
+                .min_by(|x, y| x.get_fitness().partial_cmp(&y.get_fitness()).unwrap())
+                .unwrap(),
+            pretty_config.clone(),
+        )
+        .unwrap();
+        let mean =
+            to_string_pretty(&population[population.len() / 2], pretty_config.clone()).unwrap();
+
+        let mut buffer =
+            File::create(format!("results_generation_{}.ron", generation_count)).unwrap();
+        buffer.write(b"Best: ").unwrap();
+        buffer.write(best.as_bytes()).unwrap();
+        buffer.write(b"\n").unwrap();
+        buffer.write(b"Worst: ").unwrap();
+        buffer.write(worst.as_bytes()).unwrap();
+        buffer.write(b"\n").unwrap();
+        buffer.write(b"Mean: ").unwrap();
+        buffer.write(mean.as_bytes()).unwrap();
+    }
```

Comment:
You could wrap this into a `BufWriter` to make it more performant! https://doc.rust-lang.org/stable/std/io/struct.BufWriter.html